src/preact/services/document.js

```diff
@@ -0,0 +1,54 @@
+import {urlService} from './url';
+
+class DocumentService {
+  /**
+   * @param {string} metaName
+   * @return {string|null}
+   */
+  getMetaByName(metaName) {
+    const metas = self.window.document.head.querySelectorAll('meta[name]');
+    for (let i = 0; i < metas.length; i++) {
+      const meta = metas[i];
+      const name = meta.getAttribute('name');
+      if (name === metaName) {
+        const content = meta.getAttribute('content');
+        if (content !== undefined) {
+          return content;
+        }
+      }
+    }
+    return null;
+  }
+}
+
+// eslint-disable-next-line local/no-export-side-effect
+export const docService = new DocumentService();
+
+class DocumentInfoService {
+  /**
+   * @return {string}
+   */
+  get sourceUrl() {
+    return self.window.location.href;
+  }
+
+  /**
+   * @return {string}
+   */
+  get canonicalUrl() {
+    const rootNode = self.window.document;
+
+    let canonicalUrl = rootNode?.AMP?.canonicalUrl;
+    if (!canonicalUrl) {
+      const canonicalTag = rootNode.querySelector('link[rel=canonical]');
+      canonicalUrl = canonicalTag
+        ? urlService.parse(canonicalTag.href).href
+        : this.sourceUrl;
+    }
+
+    return canonicalUrl;
+  }
+}
+
+// eslint-disable-next-line local/no-export-side-effect
```

Comment:
These lines will make it impossible to DCE any of these services. Service provider pattern instead of instantiating services directly could help.

Reply:
I agree with these concerns, and I think this is why they don't exist yet--because we haven't landed on a consensus. This feels like something that can become a really critical portion of Bento functionality, so maybe it would be good for us to have a design review or discussion thread to nail down how we want to approach this.

Reply:
@jridgewell do you have thoughts on this?

Reply:
> These lines will make it impossible to DCE any of these services

We can make them regular objects, instead of classes that need to be constructed. That will make them entirely DCE'able, without adding in a DI system.

> I've been worried that, in porting Bento components, we might miss important details

As for the full copy-paste, I think we need to reevaluate what makes sense in Bento context. A lot of this code only makes sense for AMP, including basically all of the code in this file.

Reply:
One important thing is that often the `Preact` and standard Web Component distributions do not need all the same functionality as the AMP Component.

For example, `amp-render` will use the default builtin [fetch](https://developer.mozilla.org/en-US/docs/Web/API/fetch) function for WC and Preact, only using `xhr` when for the AMP-layer. It passes the needed functionality via props:

https://github.com/ampproject/amphtml/blob/584f5a6258b2597b1ee6ad317de42d2e7268a525/extensions/amp-render/1.0/amp-render.js#L273-L277.

One thing I'm confident of us that we do not want to wholesale copy/paste Services into the `preact` directory. I'll read through this PR closely tomorrow to try and come up with a suggestion.